### ISSUE TITLE
Example of using "ETAG" and "IF-None-MATCH" tags 

### ISSUE THREAD

**Issue by jas-singh14** [eclipse/californium#664]
We have a requirement of providing support for ETAG and IF-NONE-MATCH tags in Cf server. 
Please share some sample Cf reference code which we can help us in understanding how these tags can be used.

**Comment by vikram919** [comment]
> We have a requirement of providing support for ETAG and IF-NONE-MATCH tags in Cf server.

**1. ETAG:** Purpose: Used to differentiate between representations of same resource over time.
How to use it : option is either used by clients in GET request or by server maintaining resource in response.
Server side: As per RFC 7252, https://tools.ietf.org/html/rfc7252#section-5.10.6 
> It is generated by the server providing the resource, which may generate it in any number of ways including a version, checksum, hash, or time. 

IIUC, it means server will include a meta data information as ETAG "Entity TAG".
```java
CoapServer server = new CoapServer();
server.add(new CoapResource("hello") {
      @Override
	public void handleGET(CoapExchange exchange) {
		byte[] payaload = bytes (current representation of resource);
               // tag = byte[] filled with (version, checksum, hash/time );
                exchange.setETag(tag);
		exchange.respond(code, payload);
	}
});
```
**2. If-None-Match:**: As per RFC 7252, https://tools.ietf.org/html/rfc7252#section-5.10.8.2,

> If-None-Match is useful for resource creation requests, such as PUT requests, as a means for
   protecting against accidental overwrites when multiple clients are acting in parallel on the same resource

```java
CoapServer server = new CoapServer();
server.add(new CoapResource("hello") {
  @Override
     public void handlePUT(CoapExchange exchange) {
	if(exchange.getRequestOptions().hasIfNoneMatch()) {
		OptionSet optionSet = exchange.getRequestOptions();
		if(optionSet.containsETag(etag)) {
		       exchange.respond(ResponseCode.CHANGED);
		 } else {
		       exchange.reject();
		 }
	 }
    }
});
```
PS: this is a pseudo code from the java doc provided by Californium, feel free to point my mistake

**Comment by jas-singh14** [comment]
Thanks for the information. 
Regarding#2, can you give an example how If-None-Match will be used on server side in case of GET and Post requests?

**Comment by vikram919** [comment]
> Regarding#2, can you give an example how If-None-Match will be used on server side in case of GET and Post requests?

As per RFC 7252, https://tools.ietf.org/html/rfc7252#section-5.10.8.2

If-None-Match is used with only resource creation requests such as PUT

**Comment by jas-singh14** [comment]
Seems it is possible for GET as well. https://tools.ietf.org/html/rfc7252#section-10.2.2 
Here it is mentioned that It can be used for GET method as well. So it seems it should be possible to use if-none-match for both PUT and GET. 
Considering this can you provide some reference code for GET also?
Thanks in advance..

**Comment by vikram919** [comment]
> Seems it is possible for GET as well.

As per RFC 7252, https://tools.ietf.org/html/rfc7252#section-10.2.2

> Conditional GETs:  Conditional HTTP GET requests that include an "If-
      Match" or "If-None-Match" request-header field can be mapped to a
      corresponding CoAP request.

It says HTTP GET requests that include "If-None-Match", but did not explain that CoAP request should include If-None-Match option thus it is not implemented in Californium.
Since it is not mentioned any where in standard that mapped HTTP GET requests to CoAP GET requests should include If-None-Match option as well, it only say mapped to a corresponding GET request.
 
If you find any where in standard that GET requests should/must include If-None-Match, please feel free to create a new issue specifying the line that supports your argument.

**Comment by jas-singh14** [comment]
Since raising issue and getting the resolution will take time, can you suggest some workaround (temporary fix) whereby we can meet our urgent requirement (e.g. may be by using customer Header named as if-none-match etc.) as a near term solution. Later we can definitely raise a proper issue request as well.
Please advise ASAP.

**Comment by vikram919** [comment]
`client side:`
```java
CoapClient client = new CoapClient("coap://localhost:5683/light");
//client.putIfNoneMatch("hello", MediaTypeRegistry.TEXT_PLAIN);
Request request = new Request(Code.GET);
request.getOptions().setIfNoneMatch(true);
// etag of last received data
request.getOptions().addETag(etag);
client.advanced(request);
```

**Comment by jas-singh14** [comment]
Thanks, but can you provide it for server side as well?

**Comment by vikram919** [comment]
I am not sure whether this is the right way to do so, may be @boaks will comment on it how to use it

**Comment by boaks** [comment]
@jas-singh14 

do you use the http-coap-proxy? 

If not, then why do you want to used the CoAP option "If-none-match" for a GET? For me this doesn't make sense.

> we can meet our urgent requirement (e.g. may be by using customer Header named as if-none-match etc.) as a near term solution.

"urgent requirement" and "may be by using customer Header named as if-none-match etc." seems to be the same issue as GET something only if it's not available (it's a self antithesis).

**Comment by jas-singh14** [comment]
No, we are not using http-coap-proxy..

**Comment by boaks** [comment]
> No, we are not using http-coap-proxy.

Then I wouldn't pay too much attention to 10.2.2. (which defines that http-coap-proxy).
If you really want answers according that, please ask the IETF core mailing list.

**Comment by jas-singh14** [comment]
@boaks 
Two points here:
1. Please refer to the link, https://www.logicbig.com/quick-info/web/etag-header.html which indicates that if-none-match can be used in resource retrieve request(which I think can be GET request). Is my interpretation correct?

2. Regarding your observation w.r.t. http-coap-proxy, do you mean that it might be possible for HTTP and not for CoAP. Is this understanding correct?

**Comment by boaks** [comment]
And sorry, it's really completely useless to cite **http** stuff, if there is a **coap** specific specification!

1. the link is still for **http**, but californium is **coap** :-). 

**coap** slightly change the meaning of some options. So, please read
https://tools.ietf.org/html/rfc7252#section-5.10 to understand the usage of the coap options.
For me, the "http if-none-match with etags" is similar to "coap etag used for request", see
https://tools.ietf.org/html/rfc7252#section-5.10.6.2
And for "If-none-match", which has in **coap** no additional parameter, see
https://tools.ietf.org/html/rfc7252#section-5.10.8.2

2. sure HTTP options are possible in HTTP. 

And "http If-none-match with etags" is not "coap if-none-match with etags", it's similar to "coap etags". So again, it's worth to read the specification of the coap options., if you want to use **coap**.

Just in advance: coap also changes slightly the response codes.

And, very important:
Californium provides some options (especially etag, if-match, if-none-match) only to your implementation of a `CoapResource`. Your implementation is then intended to handle that proper. From my experience, coap is perfect designed to send small request and get small responses (not saying, that some larger stuff isn't working). So in many cases a UDP datagram with the response "2.03 valid", instead of a couple of bytes with the content of the resource, is not that big advantage.

**Comment by vikram919** [comment]
As @boaks pointed out https://tools.ietf.org/html/rfc7252#section-5.10.8.2 explain that If-None-Match carries no value it means using etag with if-None-Match is completely against spec
If you want to use etag value, If-Match is used and preffered way is with only PUT requests as defined in https://tools.ietf.org/html/rfc7252#section-5.10.8.1

**Comment by boaks** [comment]
> etag with if-None-Match is completely against spec

And, as I wrote above, using etag (without the "if-blabla") in the request, should be very similar to the "http if-none-match" stuff. 
https://tools.ietf.org/html/rfc7252#section-5.10.6.2

**Comment by vikram919** [comment]
@boaks 
IIUC

1. https://tools.ietf.org/html/rfc7252#section-5.10.8.2 says:
> The If-None-Match Option carries no value. 

well, using etag+If-None-Match is not preferred way and at same time the spec also did not say it cannot/must not use etag+If-None-Match at least not even a note?

**Http2CoAP Proxy scenario:**
2. https://tools.ietf.org/html/rfc7252#section-10.2.2 spec says 
> Conditional HTTP GET requests that include an "If-Match" or "If-None-Match" request-header field can be mapped to a corresponding CoAP request.

This is also not clear again, whether it is allowed to use etag+If-None-Match for the mapped corresponding CoAP request.

> And, as I wrote above, using etag (without the "if-blabla") in the request, should be very similar to the "http if-none-match" stuff.

I would argue on this point, How does a mapped CoAP request differentiate itself with HTTP+Etag+If-None-Match to HTTP+Etag+If-Match? because both requests carries opposite responses?

**Comment by boaks** [comment]
@vikram919 

1. your in a deep-brain-dead-end :-). What http wants to achieve with "if-none-match etags" is achieved by coap simply using "etags"! so just forget the "if-none-match" stuff and provide your etags!

2. I don't care too much about that. If someone really will use it, then someone may ask the ietf core mailing list for clarification.

> I would argue on this point, How does a mapped CoAP request differentiate itself with HTTP+Etag+If-None-Match to HTTP+Etag+If-Match? because both requests carries opposite responses?

The "if-blabla" express, that I was forced to repeating over and over writing the same.
As written in https://github.com/eclipse/californium/issues/664#issuecomment-398389570
only "http if-none-match" maps to "coap etags".

**Comment by vikram919** [comment]
> your in a deep-brain-dead-end :-).

I don't think you understood my point, well I leave it your mediocrity :-)

**Comment by boaks** [comment]
It's even much harder, if there are two RFCs and the comments mix the terms!

My initial strategy was therefore trying to reduce the issue back to be only a coap issue, see
https://github.com/eclipse/californium/issues/664#issuecomment-397969932 . But with the citation of http stuff  again in https://github.com/eclipse/californium/issues/664#issuecomment-398217211 , I tried to introduce "http if-blabla" and "coap if-blabla" or "coap eatgs" to distinguish between the variants.

> well, using etag+If-None-Match is not preferred way and at same time the spec also did not say it cannot/must not use etag+If-None-Match at least not even a note?

That continues, not to distinguish between http and coap, and therefore, sure, I have no chance to understand it. For http, "http if-none-match etags GET" makes sense, and for http it is preferred! For coap, the similar function is achieved by "coap etags". And for coap "coap if-none-match" has no etag parameter and is therefore not valid.

https://tools.ietf.org/html/rfc7252#section-10.2.2
>  Conditional HTTP GET requests that include an "If-Match" or "If-None-Match" request-header field can be mapped to a corresponding CoAP request.  

Is simply unclear, and so the ietf core mailing list is the right forum to clarify it. 
It doesn't explicitly specify, that "http options" are mapped to there corresponding "coap options".
It only specifies, that a http GET request could be mapped to a corresponding coap GET request, even if "http if-none-match" or "http if-match" is used. And therefore I recommend to translate "http if-none-match etags" into "coap etags". Especially, if the http-coap-proxy isn't used at all, this answer should be enough. 
And still, if the http-coap-proxy should be used, that the clarification will be required!
(And yes, the current http-coap-proxy implementation will create wrong coap requests, though it maps the "http options" into "coap options". But, it seems to be not too frequently used and so without someone, who is willing to work on the clarification with ietf core mailing list, I will not start to fix it :-) ).

**Comment by jas-singh14** [comment]
So, if i understand correctly:
The CoAP standard should only have ETAGS and no "IF-None-Match" option for a GET request. The server will just compare the ETAG and in case it has same representation, it will return "VALID" code (but not content), otherwise it would return "CHANGED" with the current resource representation.

But at the same time in case some COAP client and server is working with both ETAG and IF-NONE-MATCH options, we can't say that those are not conforming to RFC 7252.

So implementations are free to use either way as standard doesn't explicitly defines this behaviour. So in that case, server should be able to handle both the scenarios: with if-none-match and without if-none-match.
Is my understanding correct?

Also excuse my non-technical language :-)

**Comment by boaks** [comment]
> The CoAP standard should only have ETAGS and no "IF-None-Match" option for a GET request. The server will just compare the ETAG and in case it has same representation, it will return "VALID" code (but not content), otherwise it would return "CHANGED" with the current resource representation. 

https://tools.ietf.org/html/rfc7252#section-5.10.6.2

So mostly yes, but no for the "CHANGED", in coap it's  "CONTENT".
As I also already written, not only the options have slightly changes, also the response codes.
https://github.com/eclipse/californium/issues/664#issuecomment-398389570

> But at the same time in case some COAP client and server is working with both ETAG and IF-NONE-MATCH options, we can't say that those are not conforming to RFC 7252.

Depends on the scope of the "same time".

If the scope is the same request, then I would say, it doesn't make sense to use them together, because
etags are used, if the resource is intended to be existing,
if-none-match is used, if the resource is intended to be not existing.
(But may be, someone knows a useful combination, then this will also OK for me.)

If the scope is larger and covers requests with etags and other requests with if-none-match then you can use them in that scope together without any trouble.

> So implementations are free to use either way as standard doesn't explicitly defines this behaviour. So in that case, server should be able to handle both the scenarios: with if-none-match and without if-none-match.
> Is my understanding correct?
 
No, the implementation is intended to support both, in different requests!

**Comment by jas-singh14** [comment]
@vikram919 and @boaks 

Thanks for your inputs.
